common/src/stack/pylib/stack/kvm.py

```diff
@@ -0,0 +1,323 @@
+			raise VmException(f'Failed to close hypervisor connection to {self.hostname}')
+
+	def guests(self):
+		"""
+		Return a dict of all guests defined on the
+		current hypervisor along with their current
+		status.
+
+		Raises a VmException if the guest list could
+		not be retrieved
+		"""
+
+		vm_guests = {}
+
+		try:
+			domains = self.kvm.listAllDomains()
+
+			except libvirtError as msg:
+				raise VmException(f'Failed to get list of VM domains on {self.hostname}:\n{msg}')
+
+		for dom in domains:
+			status = 'off'
+			if dom.isActive() == True:
+				status = 'on'
+			vm_guests[dom.name()] = status
+
+		return vm_guests
+
+	def add_domain(self, dom_config):
+		"""
+		Define a new guest in the hypervisor
+
+		Raises a VmException if the failed to
+		be added
+		"""
+
+		try:
+			self.kvm.defineXML(dom_config)
+		except libvirtError as msg:
+
+			# If a vm is already defined do nothing
+			# but raise an exception for other errors
+			if 'already exists with' in str(msg):
+				return
+			raise VmException(f'Failed to define guest on hypervisor {self.hostname}:\n{msg}')
+
+	def remove_domain(self, guest_name):
+		"""
+		Remove a guest domain with the given name
+
+		Raises a VmException if the VM could not be
+		removed
+		"""
+
+		try:
+			dom = self.kvm.lookupByName(guest_name)
+			dom.undefine()
+		except libvirtError as msg:
+			raise VmException(f'Failed to undefine VM {guest_name} on hypervisor {self.hostname}:\n{str(msg)}')
+
+	def start_domain(self, guest_name):
+		"""
+		Start a domain on a hypervisor
+
+		Raises a VmException if the VM
+		failed to be started
+		"""
+		try:
+			dom = self.kvm.lookupByName(guest_name)
+			dom.create()
+		except libvirtError as msg:
+			raise VmException(f'Failed to start VM {guest_name} on hypervisor {self.hostname}:\n{msg}')
+
+	def stop_domain(self, guest_name):
+		"""
+		Stop a domain on a hypervisor
+
+		Raises a VmException if the VM could
+		not be stopped
+		"""
+
+		try:
+			dom = self.kvm.lookupByName(guest_name)
+
+                        # This sounds scary but just stops the vm
+			dom.destroy()
+		except libvirtError as msg:
+			raise VmException(f'Failed to stop VM {guest_name} on hypervisor {self.hostname}:\n{msg}')
+
+	def add_pool(self, poolname, pooldir):
+		"""
+		Define a new kvm storage pool
+		with the given pool directory and name
+
+		Returns empty string if the pool exists
+		and the pool name if the pool was created
+
+		Raises a VmException if the pool failed
+		to be created
+		"""
+
+		poolxml = f"""<pool type="dir">
+	<name>{poolname}</name>
+	<target>
+		<path>{pooldir}</path>
+	</target>
+	</pool>"""
+
+		# Need to ensure the directory exists on the kvm host
+		self.do_remote_command('root', [ f'mkdir -p {pooldir}' ])
+		try:
+			pools = self.kvm.listAllStoragePools(0)
+			pool_exists = self.kvm.storagePoolLookupByName(poolname)
+			if pool_exists:
+				return
+			pool = self.kvm.storagePoolDefineXML(poolxml, 0)
+			pool.create()
+			pool.setAutostart(1)
```

Comment:
Please document in the docstring that this sets the storage pool to autostart, and what that potentially means for users of this function.